jib-gradle-plugin/src/main/java/com/google/cloud/tools/jib/gradle/TargetImageParameters.java

```diff
@@ -34,42 +36,51 @@
 
   private final AuthParameters auth;
 
-  @Nullable private String image;
-  private Set<String> tags = Collections.emptySet();
+  private Property<String> image;
+  private SetProperty<String> tags;
   @Nullable private String credHelper;
 
   @Inject
   public TargetImageParameters(ObjectFactory objectFactory) {
     auth = objectFactory.newInstance(AuthParameters.class, "to.auth");
+    image = objectFactory.property(String.class);
+    tags = objectFactory.setProperty(String.class);
   }
 
   @Input
-  @Nullable
-  @Optional
   public String getImage() {
     if (System.getProperty(PropertyNames.TO_IMAGE) != null) {
       return System.getProperty(PropertyNames.TO_IMAGE);
     }
-    return image;
+    return image.getOrNull();
   }
 
   public void setImage(String image) {
-    this.image = image;
+    this.image.set(image);
+  }
+
+  public void setImage(Provider<String> image) {
+    if (image != null) {
```

Comment:
I speculate that we won't need this null check.

Reply:
hm removing this check is resulting in an `UnexpectedBuildFailure` with the message, `Cannot set the value of a property using a null provider` in places where the target image value is not set. (For example, https://github.com/GoogleContainerTools/jib/blob/f892bdba90cbe64a151e0196fa0c8e633f529a7f/jib-gradle-plugin/src/test/java/com/google/cloud/tools/jib/gradle/skaffold/FilesTaskV2Test.java#L110

Reply:
Aha, looks like Gradle prefers this method signature over `String`. If that's the case, I think it should be

```suggestion
    if (image == null) {
      this.image = null;
    } else {
      this.image.set(image);
    }
```

Reply:
Ah, exactly! You're right, when `jib.to.image` is set to `null` then Gradle prefers this method. With the `if (image != null)` checker, we are currently returning the message `Missing target image parameter, perhaps you should add a 'jib.to.image' configuration parameter to your build.gradle...` if `jib.to.image` is not specified, which seems like expected behavior. I could be missing something here but why do we want to make the Property be null instead of not just not setting a value for it?

Reply:
Good question.

From the general API perspective, irrelevant of Jib: let's say a Java developer is using some Java library which has `Person.setEmail(String)`. It's an optional property, and by default `getAdress()` will return `null`. Let's say the developer does
   ```java
   // prints out "no email given"
   System.out.println(person.getEmail() == null ? "no email given" : person.getEmail());
   person.setEmail("foo@example.com");
   person.setEmail(null);
   System.out.println(person.getEmail() == null ? "no email given" : person.getEmail());
   ```
   Then, which would be reasonable for the last `println()` to print: `no email given` or `foo@example.com`? Of course, it may depend on the situation, but in this case, I expect `setEmail(null)` to clear the previous value and make the property "missing"; it is counter-intuitive it becomes a no-op. 

Also note that `jib.to.image` is basically an optional property (as indicated by the `@Optional` annotation). Of course, it is required when pushing to a remote registry (`jib:build`), and it will say 'Missing target image parameter'. However, it isn't required when pushing to a local Docker daemon (`jib:dockerBuild`); in this case, Jib will auto-pick some name for it.

But, before we add the null check, I'd like to verify Gradle always calls `setImage(null)` even when `jib.to.image` is simply missing (not configured) rather than when `jib.to.image` is explicitly set to `null`. Reading https://github.com/GoogleContainerTools/jib/pull/2739/files#r485907745, it looks like this is only when explicitly setting `jib.to.image` to `null`. In that case, I don't think we need the null check actually; I think we can just fix the failing test.

Reply:
Thank you so much for the detailed explanation! 

And with regards to the last point about https://github.com/GoogleContainerTools/jib/pull/2739/files#r485907745, the issue only occurs when `jib.to.image` is explicitly set to `null`. I believe the tests that were failing set `jib.to.image` to `System.getProperty("_TARGET_IMAGE") and didn't initialize `_TARGET_IMAGE`? Because System.getProperty(String key) returns `null` if there is no property with that key.

Reply:
Yeah, that makes sense. I think we don't actually need a null check; we can just fix the test, e.g., setting `_TARGET_IMAGE` to "ignored."